client/src/components/Popup.tsx

```diff
@@ -0,0 +1,32 @@
+import { ReactNode } from 'react';
+import { createPortal } from 'react-dom';
+
+type Props = {
+  children: ReactNode;
+  isOpen: boolean;
+  buttonRef?: React.RefObject<HTMLButtonElement> | null;
+  onShadeClick: () => void;
+};
+export function Popup({ children, isOpen, buttonRef, onShadeClick }: Props) {
+  const buttonRectangle = buttonRef?.current?.getBoundingClientRect();
+  const { innerWidth } = window;
+
+  const top = buttonRectangle ? buttonRectangle.bottom : '50%';
+  const right = buttonRectangle ? innerWidth - buttonRectangle.right : '50%';
+
+  if (isOpen) {
```

Comment:
If `isOpen` is false, you return `undefined`, because there's no return statement. I think code is better written if there is always an explicit return statement (unless it's a void function). So I think a "fail fast" would be better here. Probably the first line of the function should be something like, `if (!isOpen) return null;`.